adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -496,6 +504,10 @@ private void doUploadMessage(final RoutingContext ctx, final String tenant, fina
                 final Future<JsonObject> tokenTracker = getRegistrationAssertion(tenant, deviceId, authenticatedDevice);
                 final Future<TenantObject> tenantConfigTracker = getTenantConfiguration(tenant);
 
+                final AtomicBoolean downstreamMessageSent = new AtomicBoolean(false);
```

Comment:
The code reads not very fluent with this marker, which is set and checked deep inside different code blocks. Maybe we can find a better solution in a future version.

Reply:
Full ack to this - this serves as the now working version which is already cleaned up compared to the previous prototype, but still needs some atomic `helper` objects.

I am sure in the first milestone of 0.7 there will be an improved, better readable solution, but for now this is what we have here and what works in all cases:

- the command is received **before** the downstream message was sent
- the command is received completely parallel while the downstream message is being sent
- the command is received strictly **after** the downstream message was sent to the AMQP network

So, I propose to keep it for 0.6 like this and come up with a better readable version afterwards.